src/python/pants/source/source_root.py

```diff
@@ -361,6 +366,25 @@ def _do_add_pattern(self, pattern, langs, category):
     node.category = category
     node.is_terminal = True
 
+  def traverse(self) -> Set[UncanonicalizedSourceRoot]:
+    uncanonicalized_source_roots = set()
+    lang_canonicalizations = self._source_root_factory._lang_canonicalizations
+    all_lang_names = tuple(lang_canonicalizations.keys())
+
+    def traverse_helper(node, path_components):
```

Comment:
Type hints encouraged for inner functions. Not required and no one will block if you don't have it, but if you already know the types would be great to add :)